specs/chugsplash-config.md

```diff
@@ -0,0 +1,276 @@
+
+## Contract Definition
+JSON or Javascript object defining a contract to be deployed, it's constructor args, and variable definitions. Contract definitions must conform to the following format:
+
+```js
+ReferenceName: {
+  contract: 'LiteralContractName',
+  constructorArgs: {},
+  variables: {}
+}
+```
+
+## Contract Constructor Args
+A JSON or Javascript object defining a set of constructor args. All constructor arguements are required to be immutable variables...
+
+TODO finish defining constructor args spec
+
+## Contract Variables
+A JSON or Javascript object defining a set of ChugSplash input variables. ChugSplash is designed to remove the need for constructors or initializers by allowing the user to define the value of their variables directly during the deployment process. To ensure this process works reliably we perform input validation during the bundling process. Below we define the expected valid types. All other input types should be caught and rejected.
+
+Note that ChugSplash config files may be written in Javascript and therefore users are able to take advantage of various Javscript features and libraries to fetch values to be used in their ChugSplash config files.
+
+## uint & int
+For uints and ints, we allow two valid input types:
+- Base 10 string numbers: "1", "10", "1000000000000000000", "-100"
+- Base 10 literal numbers: 1, 10, 100, -100
+
+We enforce valid input ranges based on the specific byte size and sign of the target solidity variable. For example, if a variable is defined as uint8, we expect that values outside the range 0:255 to be caught and rejected during our bundling process. Likewise for int8, we expect that values outside the range -127:128 would be caught and rejected.
+
+Note that we explicitly do not support alternative base inputs such as hexadecimal for uint and int.
+
+### Examples
+```js
+// Solidity
+uint public x;
+
+// Javascript
+x: "1",
+```
+
+```js
+// Solidity
+uint public x;
+
+// Javascript
+x: 1,
+```
+
+```js
+// Solidity
+int public x;
+
+// Javascript
+x: -1,
+```
+
+```js
+// Solidity
+int public x;
+
+// Javascript
+x: "-1",
+```
+
+## Bytes32
+For bytes32, we allow only exactly 32 byte hex strings.
+
+### Examples
+```js
+// Solidity
+bytes32 public x;
+bytes32 public y;
+
+// Javascript
+x: '0x1111111111111111111111111111111111111111111111111111111111111111',
+y: ethers.utils.hexZeroPad(ethers.utils.hexlify(1), 32),
+```
+
+## Strings
+For strings, we allow only string literals.
+Note that if you supply a hex value in string format, we will not interpret it as a bytes value. We will encode it as if it were any other string.
+
+### Example
+```js
+// Solidity
+string public x;
+
+// Javascript
+x: 'literalString',
+```
+
+## Booleans
+For booleans, we allow literal Javascript or JSON booleans or the literal strings "true" and "false".
```

Comment:
Thoughts on just allowing `true` and `false` instead of their string equivalents? Less if-statements down the road during testing and writing the decoding logic

Reply:
Hmm, good question. In my experience, it's pretty standard for libraries to support the literal strings "true" and "false". However, in our case, I can see an argument that it could introduce longer-term bugs. I'm cool with removing this feature.